spec/lib/s3_cors_rulesets_spec.rb

```diff
@@ -0,0 +1,209 @@
+        expect(result[:backup_rules_applied]).to eq(true)
+      end
+
+      it "applies the ASSETS rules and the DIRECT_UPLOAD rules when S3 direct uploads are enabled" do
+        SiteSetting.enable_direct_s3_uploads = true
+
+        client.stub_responses(:get_bucket_cors, {})
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::ASSETS
+            ]
+          }
+        )
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::DIRECT_UPLOAD
+            ]
+          }
+        )
+        result = sync_rules
+        expect(result[:assets_rules_applied]).to eq(true)
+        expect(result[:direct_upload_rules_applied]).to eq(true)
+        expect(result[:backup_rules_applied]).to eq(false)
+      end
+
+      it "applies all rules if everything is set up" do
+        SiteSetting.enable_direct_s3_uploads = true
+        setup_backups
+
+        client.stub_responses(:get_bucket_cors, {})
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::ASSETS
+            ]
+          }
+        )
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::DIRECT_UPLOAD
+            ]
+          }
+        )
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-backup-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::BACKUP_DIRECT_UPLOAD
+            ]
+          }
+        )
+        result = sync_rules
+        expect(result[:assets_rules_applied]).to eq(true)
+        expect(result[:direct_upload_rules_applied]).to eq(true)
+        expect(result[:backup_rules_applied]).to eq(true)
+      end
+
+      it "does no changes if all the rules already exist" do
+        SiteSetting.enable_direct_s3_uploads = true
+        setup_backups
+
+        client.stub_responses(:get_bucket_cors, {
+          cors_rules: [S3CorsRulesets::ASSETS, S3CorsRulesets::BACKUP_DIRECT_UPLOAD, S3CorsRulesets::DIRECT_UPLOAD]
+        })
+        client.expects(:put_bucket_cors).never
+        result = sync_rules
+        expect(result[:assets_rules_applied]).to eq(false)
+        expect(result[:direct_upload_rules_applied]).to eq(false)
+        expect(result[:backup_rules_applied]).to eq(false)
+      end
+
+      def setup_backups
+        SiteSetting.enable_backups = true
+        SiteSetting.s3_backup_bucket = "s3-backup-upload-bucket"
+        SiteSetting.s3_use_iam_profile = true
+        SiteSetting.backup_location = BackupLocationSiteSetting::S3
+      end
```

Comment:
Hmm do we need this when the global settings have already been configured?

Reply:
Yes we do, see:

https://github.com/discourse/discourse/blob/fada309c9a5c70c2c757a09b001ecfd9b568422f/lib/site_settings/validations.rb#L159-L167

I agree it is a little strange, confused me a bit why that was needed but the backup bucket stuff is not loaded into GlobalSetting